apiserver/application/application.go

```diff
@@ -945,6 +945,44 @@ func (api *API) saveRemoteApplication(
 			Scope:     ep.Scope,
 		}
 	}
+
+	// If the a remote application with the same name and endpoints from the same
+	// source model already exists, we will use that one.
+	existingRemoteApp, err := api.backend.RemoteApplication(applicationName)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, errors.Trace(err)
+	}
+	// We may be able to re-use the existing remote app if it is from the same
+	// model and the endpoints are compatible.
+	if err == nil && existingRemoteApp.SourceModel().Id() == sourceModelTag.Id() {
+		newEpsMap := make(map[charm.Relation]bool)
+		for _, ep := range remoteEps {
+			newEpsMap[ep] = true
+		}
+		existingEps, err := existingRemoteApp.Endpoints()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		maybeSameEndpoints := len(newEpsMap) == len(existingEps)
+		for _, ep := range existingEps {
+			delete(newEpsMap, ep.Relation)
+		}
+		sameEndpoints := maybeSameEndpoints && len(newEpsMap) == 0
+		if sameEndpoints {
+			return existingRemoteApp, nil
+		}
+
+		// Update the existing remote app to have the new, additional endpoints.
```

Comment:
maybe expand the comment, noting that AddEndpoints will return an error if any existing endpoints have been changed. it's not clear from the comments that we're intentionally passing along the existing-but-different ones

I think it might be clearer if check they're equal here, and don't pass them along; and have AddEndpoints complain if the endpoint already exists